### PR TITLE
feat(files): publish task for Files API as content storage

### PR DESCRIPTION
**Pull request by iamfrontender**
https://evrythng.atlassian.net/browse/FE-45

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -126,8 +128,15 @@ You can use any of these services to host your bundles:
 * [AWS S3](https://aws.amazon.com/s3/)
 * [Google Cloud Platform](https://cloud.google.com/)
 * [Netlify](https://www.netlify.com/)
+* [Files API](#deploying-using-files-api)
 * ...
 
+### Deploying using Files API
+
+EVRYTHNG platform provides powerful cloud file system, or Files API, which could be used
```
Comment (C-D-Lewis) [inline README.md]: "The EVRYTHNG Platform provides a powerful cloud file storage system..."

File: README.md [hunk 2]
```diff
@@ -126,8 +128,15 @@ You can use any of these services to host your bundles:
...
+to store your custom components. Files API requires an API key with valid permissions to be provided from environment. It could be done either with `EVT_AUTH` environment variable or `~./evt-auth` file.
```
Comment (C-D-Lewis) [inline README.md]: "The Files API requires..."

File: README.md [hunk 3]
```diff
@@ -93,6 +93,8 @@ Here's a list of available tasks:
   * scaffolds a new component. [Read below](#generating-components) for usage details.
 * `service`
   * scaffolds a new service. [Read below](#generating-components) for usage details.
+* `publish`
+  * builds and deploys the components to EVRYTHNG platform using Files API. [Read below](#deploying-using-files-api) for usage details.
```
Comment (joaovieira) [inline README.md]: the Files API. provide link to Files API documentation in readme.io

File: README.md [hunk 4]
```diff
@@ -126,8 +128,15 @@ You can use any of these services to host your bundles:
 * [AWS S3](https://aws.amazon.com/s3/)
 * [Google Cloud Platform](https://cloud.google.com/)
 * [Netlify](https://www.netlify.com/)
+* [Files API](#deploying-using-files-api)
 * ...
 
+### Deploying using Files API
```
Comment (joaovieira) [inline README.md]: the Files API

File: gulpfile.babel.js [hunk 5]
```diff
@@ -1,60 +1,27 @@
...
-      .join('');
-  };
-  const name = yargs.argv.name;
-  const parentPath = yargs.argv.parent || '';
-  const destPath = path.join(root, type + 's', parentPath, name);
+/**
+ * Bundles application modules with webpack
+ */
+gulp.task('webpack', done => {
+  webpackConfig.entry[config.name] = config.paths.entry;
```
Comment (joaovieira) [inline gulpfile.babel.js]: If you want to make the name configurable, it should be the same for the 'webpack' and the 'serve' tasks (serve is still using 'components'). Also, they can be defined directly in the webpack.config.js.

Comment (joaovieira) [inline gulpfile.babel.js]: `webpack` is the build task, should use the dist config. `webpack` and `serve` are swapped.

File: gulpfile.babel.js [hunk 6]
```diff
@@ -65,17 +32,18 @@ gulp.task('webpack', (cb) => {
...
+ * Starts webpack dev server
+ */
 gulp.task('serve', () => {
-  var config = require('./webpack.config');
-  config.entry.components = paths.entry;
+  let compiler;
 
-  var compiler = webpack(config);
+  webpackDistConfig.entry.components = config.paths.entry;
+  compiler = webpack(webpackDistConfig);
```
Comment (joaovieira) [inline gulpfile.babel.js]: `const compiler`

File: gulpfile.babel.js [hunk 7]
```diff
@@ -89,15 +57,27 @@ gulp.task('serve', () => {
...
 
+/**
+ * Deploys new service template
+ */
 gulp.task('service', () => copyTemplate('service'));
 
+/**
+ * Uploads the built widgets to a EVRYTHNG File entity.
+ */
+gulp.task('publish', ['build'], () => deployWidgets());
```
Comment (joaovieira) [inline gulpfile.babel.js]: upgrade to Gulp 4

File: tasks/config.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+};
+
+/**
+ * Makes given glob relative to sources root
+ *
+ * @private
+ * @param {String} glob
+ * @returns {string|*}
+ */
+function inRoot(glob = '') {
```
Comment (joaovieira) [inline tasks/config.js]: `resolveToRoot` is better/more expressive (why changing existing code?)

File: tasks/copy-template.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
+'use strict';
```
Comment (joaovieira) [inline tasks/copy-template.js]: no need to use strict in ES modules, btw

File: tasks/copy-template.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+import gulp from 'gulp';
+import rename from 'gulp-rename';
+import template from 'gulp-template';
+
+import config from './config';
+
+export default (type) => {
+  const name = yargs.argv.name;
+  const parentPath = yargs.argv.parent || '';
+  const destPath = path.join(root, type + 's', parentPath, name);
```
Comment (joaovieira) [inline tasks/copy-template.js]: what's root in here? did you try this?

File: tasks/deploy-widgets.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+  file: path.join(
+    webpackConfig.output.path,
+    webpackConfig.output.filename.replace('[name]', projectConfig.name)
+  ),
+  filesFilter: {
+    tags: ['components']
+  }
+};
+
+////////////////////////////////////////////
```
Comment (joaovieira) [inline tasks/deploy-widgets.js]: remove

File: tasks/deploy-widgets.js [hunk 12]
```diff
@@ -0,0 +1,85 @@
+'use strict';
+
+import fs from 'fs';
+import path from 'path';
+import request from 'request';
+import evrythng from 'evrythng-extended';
+
+import webpackConfig from '../webpack.config';
+import getApiKey from './get-api-key';
```
Comment (joaovieira) [inline tasks/deploy-widgets.js]: modules should not contain verbs (moreover that module exports several things...). call it `evtUtils` or something like that, with a readApiKey exported method.

File: tasks/get-api-key.js [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+  env: 'EVT_AUTH',
+  file: '.evt-auth'
+};
+
+/////////////////////////////////////////////////////////////////
+
+export {
+  getAPIKey as default,
+
+  readEnvironment,
```
Comment (joaovieira) [inline tasks/get-api-key.js]: why get for some, read for others? keep it consistent.

File: tasks/deploy-widgets.js [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+import projectConfig from './config';
```
Comment (joaovieira) [inline tasks/deploy-widgets.js]: config = config

File: tasks/deploy-widgets.js [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+
+const config = {
```
Comment (joaovieira) [inline tasks/deploy-widgets.js]: yet another config. you can have multiple constant variables..

File: tasks/deploy-widgets.js [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+  file: path.join(
+    webpackConfig.output.path,
```
Comment (joaovieira) [inline tasks/deploy-widgets.js]: dist config

File: tasks/deploy-widgets.js [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+    tags: ['components']
+  }
+};
+
+////////////////////////////////////////////
+
+export default () => {
+  return getApiKey()
+    .then(apiKey => evrythng.setup({ apiKey }))
+    .then(() => console.log('//////////', evrythng.settings.apiKey))
```
Comment (joaovieira) [inline tasks/deploy-widgets.js]: is this meant to stay? a log with '////' doesn't say anything

File: tasks/deploy-widgets.js [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+    webpackConfig.output.filename.replace('[name]', projectConfig.name)
+  ),
+  filesFilter: {
+    tags: ['components']
+  }
+};
+
+////////////////////////////////////////////
+
+export default () => {
```
Comment (joaovieira) [inline tasks/deploy-widgets.js]: comment/docs

File: tasks/deploy-widgets.js [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+  return getApiKey()
+    .then(apiKey => evrythng.setup({ apiKey }))
+    .then(() => console.log('//////////', evrythng.settings.apiKey))
+    .then(() => Promise.all([getMeta(), getStats()]))
+    .then(upload);
+};
+
+function getMeta() {
```
Comment (joaovieira) [inline tasks/deploy-widgets.js:42]: comment/docs, same below

File: tasks/deploy-widgets.js [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+      name: `${projectConfig.name}.js`,
+      type: 'application/javascript',
+      tags: config.filesFilter.tags,
+      privateAccess: false
+    }
+  });
+}
+
+function upload([ meta, stats ]) {
+  return new Promise((res, rej) => {
```
Comment (joaovieira) [inline tasks/deploy-widgets.js]: resolve, reject (res may be misunderstood by response, which is the most common)

File: tasks/get-api-key.js [hunk 21]
```diff
@@ -0,0 +1,79 @@
+'use strict';
+
+import fs from 'fs';
+import path from 'path';
+import expandTilde from 'expand-tilde';
+
+let config = {
```
Comment (joaovieira) [inline tasks/get-api-key.js]: another config, use first level constants instead

File: tasks/get-api-key.js [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+  env: 'EVT_AUTH',
+  file: '.evt-auth'
+};
+
+/////////////////////////////////////////////////////////////////
```
Comment (joaovieira) [inline tasks/get-api-key.js]: remove

File: tasks/get-api-key.js [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+
+export {
```
Comment (joaovieira) [inline tasks/get-api-key.js]: use export keywords on the methods instead of exporting an object

File: tasks/get-api-key.js [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+  getAPIKey as default,
+
+  readEnvironment,
+  readFile,
+  setup
+};
+
+/////////////////////////////////////////////////////////////////
```
Comment (joaovieira) [inline tasks/get-api-key.js]: remove

File: tasks/get-api-key.js [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+
+function getAPIKey() {
```
Comment (joaovieira) [inline tasks/get-api-key.js]: comment/docs

File: README.md (hunk 2 above)
Comment (joaovieira) [reply README.md]: do not force people to dirty their home environment.

1) use env variable; or allow people to pass the key in the arguments
2) what would be the structure of the file? if there's only one variable, there's no need for a file. also, we may have a proper credentials/profiles/configuration file in the future, so we better not define such a structure now

(path was also wrong: `~/.evt-auth`)

File: tasks/get-api-key.js [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+/**
+ * Tries to read API key from authorisation file
+ *
+ * @returns {Promise}
+ */
+function readFile() {
+  return new Promise((res, rej) => {
+    let authFile = path.join(expandTilde('~'), config.file);
```
Comment (joaovieira) [inline tasks/get-api-key.js]: see comment in readme, remove this tilde dependency

File: tasks/get-api-key.js [hunk 27]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * Allows to change env variable name or other configuration
+ * options from external code.
+ *
+ * @param {Object} newConfig
+ *   @prop {string} env Environment Variable name
+ *   @prop {string} file Name of auth file
+ */
+function setup(newConfig) {
```
Comment (joaovieira) [inline tasks/get-api-key.js]: where is this used?

File: webpack.config.js [hunk 28]
```diff
@@ -1,4 +1,4 @@
-var path    = require('path');
+var path = require('path');
```
Comment (joaovieira) [inline webpack.config.js]: const or import

File: webpack.config.js [hunk 29]
```diff
@@ -1,4 +1,4 @@
...
 
 module.exports = {
```
Comment (joaovieira) [inline webpack.config.js]: es6 modules

**Review by joaovieira — CHANGES_REQUESTED** [review]

File: gulpfile.babel.js [hunk 30]
```diff
@@ -65,6 +60,11 @@ gulp.task('serve', () => {
 });
 
 /**
+ * Alias for webpack task
+ */
+gulp.task('build', gulp.series('webpack'));
```
Comment (joaovieira) [inline gulpfile.babel.js]: https://github.com/gulpjs/gulp/issues/844#issuecomment-68311119

File: tasks/copy-template.js [hunk 31]
```diff
@@ -1,5 +1,3 @@
-'use strict';
```
Comment (joaovieira) [inline tasks/copy-template.js]: remove everywhere

File: tasks/deploy-widgets.js [hunk 32]
```diff
@@ -5,64 +5,90 @@ import path from 'path';
...
-////////////////////////////////////////////
-
+/**
+ * Deploys the built components to EVRYTHNG Files API.
+ *
+ * @return {Promise}
+ */
 export default () => {
-  return getApiKey()
+  return readAPIKeyFromEnvironment()
```
Comment (joaovieira) [inline tasks/deploy-widgets.js]: environment means environment variable? no need to add implementation details to the method name (we can change the way you read it at some point in the future). simply `readApiKey` is fine

File: tasks/evt-utils.js [hunk 33]
```diff
@@ -0,0 +1,10 @@
+'use strict';
+
+/**
+ * Tries to read API key from environment variable.
+ *
+ * @returns {Promise}
+ */
+export function readAPIKeyFromEnvironment() {
+  return Promise.resolve(process.env['EVT_AUTH']);
```
Comment (joaovieira) [inline tasks/evt-utils.js]: if it's just for this, it can stay in the deploy file

File: tasks/deploy-widgets.js [hunk 34]
```diff
@@ -5,64 +5,90 @@ import path from 'path';
...
+/**
+ * Performs actual upload of file from taskConfig to remote storage.
+ *
+ * @param {Object} meta - EVRYTHNG File entity
+ * @param {Object} stats - fs.stats of the file to upload
+ * @return {Promise}
+ */
 function upload([ meta, stats ]) {
   return new Promise((res, rej) => {
     let req = request.put({
```
Comment (joaovieira) [inline tasks/deploy-widgets.js]: `const req`, `const fileStream`.

use `resolve` and `reject` as well.

**Review by joaovieira — CHANGES_REQUESTED** [review]

File: gulpfile.babel.js [hunk 35]
```diff
@@ -57,29 +57,37 @@ gulp.task('serve', () => {
...
-});
+}
 
 /**
- * Alias for webpack task
+ * Uploads the built widgets to a EVRYTHNG File entity.
  */
-gulp.task('build', gulp.series('webpack'));
+function publish() {
+  return gulp.series(compile, deployWidgets)();
```
Comment (joaovieira) [inline gulpfile.babel.js]: don't use gulp series/parallel methods in tasks definitions. use only in task registration below (that's where you orchestrate the task flow using the various subtasks):

```javascript
gulp.task('publish', gulp.series(build, publish));
```

**Review by joaovieira — CHANGES_REQUESTED** [review]

**Review by joaovieira — APPROVED** [review]

**Pull request merged**